syntax = "proto3";

package inventory;

option go_package = "inventorypb";

// ==========================
// Message Definitions
// ==========================

message Empty {}

message ProductRequest {
  int32 product_id = 1;
}

message ProductUpdateRequest {
  int32 product_id = 1;
  int32 stock_change = 2; // số lượng tăng (+) hoặc giảm (-)
}

message Product {
  int32 id = 1;
  string name = 2;
  string description = 3;
  double price = 4;
  int32 stock = 5;
  string created_at = 6;
  string updated_at = 7;
}

message ProductResponse {
  Product product = 1;
}

message UpdateInventoryResponse {
  string message = 1;
  Product updated_product = 2;
}

message InventoryList {
  repeated Product products = 1;
}

// ==========================
// Service Definitions
// ==========================

service InventoryService {
  // Lấy thông tin inventory của một sản phẩm
  rpc GetInventory(ProductRequest) returns (ProductResponse);

  // Cập nhật số lượng tồn kho (tăng/giảm)
  rpc UpdateInventory(ProductUpdateRequest) returns (UpdateInventoryResponse);

  // Lấy danh sách toàn bộ sản phẩm trong kho
  rpc ListInventory(Empty) returns (InventoryList);
}
